idgenerator/aws/xray/aws_xray_idgenerator.go

```diff
@@ -0,0 +1,77 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aws // import "go.opentelemetry.io/contrib/idgenerator/aws/xray"
+
+import (
+	crand "crypto/rand"
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"math/rand"
+	"strconv"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/otel/trace"
+)
+
+type IDGenerator struct {
+	sync.Mutex
+	randSource *rand.Rand
+}
+
+// NewIDGenerator returns an idGenerator used for sending traces to AWS X-Ray
+func NewIDGenerator() *IDGenerator {
+	gen := &IDGenerator{}
+	var rngSeed int64
+	_ = binary.Read(crand.Reader, binary.LittleEndian, &rngSeed)
+	gen.randSource = rand.New(rand.NewSource(rngSeed))
+	return gen
+}
+
+// NewSpanID returns a non-zero span ID from a randomly-chosen sequence.
+func (gen *IDGenerator) NewSpanID() trace.SpanID {
+	gen.Lock()
+	defer gen.Unlock()
+	sid := trace.SpanID{}
+	gen.randSource.Read(sid[:])
+	return sid
+}
+
+// NewTraceID returns a non-zero trace ID based on AWS X-Ray TraceID format.
+// (https://docs.aws.amazon.com/xray/latest/devguide/xray-api-sendingdata.html#xray-api-traceids)
+func (gen *IDGenerator) NewTraceID() (trace.TraceID, error) {
+	gen.Lock()
+	defer gen.Unlock()
+
+	tid := trace.TraceID{}
+	currentTime, err := getCurrentTimeHex()
+	if err != nil {
+		return trace.TraceID{}, err
+	}
+	copy(tid[:4], currentTime)
+	gen.randSource.Read(tid[4:])
+
+	return tid, nil
+}
+
+func getCurrentTimeHex() ([]uint8, error) {
+	currentTime := time.Now().Unix()
+	currentTimeHex, err := hex.DecodeString(strconv.FormatInt(currentTime, 16))
+	if err != nil {
+		return nil, errors.New("cannot convert current timestamp to hex")
```

Comment:
```suggestion
		return nil, fmt.Errorf("cannot convert current timestamp to hex: %w", err)
```

Reply:
Hi @Aneurysm9 , right now, because I am handling the error from `hex.DecodeString()`, my `NewTraceID()` function returns `(trace.TraceID, error)` instead of just a `trace.TraceID`. Wouldn't that collide with the interface you made for IDGenerator where the `NewTraceID()` function should only return a `trace.TraceID`? https://github.com/Aneurysm9/opentelemetry-go/blob/OTG-1351/sdk/trace/id_generator.go#L28

Reply:
Fair point, and I'm not sure whether `NewTraceID()` should have the possibility of erroring.  Having read the implementations of `hex.DecodeString()` and `strconv.FormatInt()` it seems reasonably safe to assume that no error should result from this operation.  Perhaps leave a comment indicating that the two error conditions for `hex.DecodeString()`, odd-length strings and non-hex digits, are not produced by `strconv.FormatInt()` so that anyone coming along later doesn't question ignoring the error.